### ISSUE TITLE
Azure build pipelines for MSVC doesn't work

### ISSUE THREAD

**Issue by redboltz** [redboltz/mqtt_cpp#712]
I noticed that resent MSVC build is not checked.
The environment no longer define `BOOST_ROOT`. But defined `BOOST_ROOT_1_72_0` and `BOOST_ROOT_1_69_0`.

I tried to fix it at #710. But I got other errors. I'm not expert of MSVC. I stuck.

**Comment by redboltz** [comment]
I remembered https://github.com/redboltz/mqtt_cpp/pull/687#issuecomment-709869027 .
I will try this.

**Comment by redboltz** [comment]
Error message has changed from

https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1563&view=logs&j=77262b04-dc69-55ba-9da0-0c8477d74276&t=1b86177c-f53d-5127-4b2f-4b26b44c9d86&l=56

to

https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1573&view=logs&j=b02611a8-afa6-57f9-a609-f526117c2f48&t=b31961eb-d0d8-59c5-9528-37bb2b45de1e&l=55

if I define `BOOST_TEST_SOURCE`.

Hmm...

**Comment by redboltz** [comment]
@kleunen you said https://github.com/redboltz/mqtt_cpp/pull/687#issuecomment-709869027 . Could you take a look the issue.
We are developing without MSVC test these days.

**Comment by kleunen** [comment]
Yes, I have also seen the inconsistent dll linkage. But if I define BOOST_TEST_SOURCE in test_main.hpp, it compiles fine for me:

````C++
// Copyright Takatoshi Kondo 2015
//
// Distributed under the Boost Software License, Version 1.0.
// (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)

#if !defined(MQTT_TEST_MAIN_HPP)
#define MQTT_TEST_MAIN_HPP

#define BOOST_TEST_SOURCE
#define BOOST_TEST_MODULE mqtt_client_cpp_test
#include <boost/test/unit_test.hpp>

#endif // MQTT_TEST_MAIN_HPP

**Comment by redboltz** [comment]
I got many 

```
C:\hostedtoolcache\windows\boost\1.69.0\x86_64\include\boost-1_69\boost/winapi/config.hpp(122,7): error C2124: divide or mod by zero [D:\a\1\s\build\test\resend.vcxproj]
```
https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1573&view=logs&j=b02611a8-afa6-57f9-a609-f526117c2f48&t=b31961eb-d0d8-59c5-9528-37bb2b45de1e&l=55

errors when I define BOOST_TEST_SOURCE.


Before I define BOOST_TEST_SOURCE, I got the following errors:

```
C:\hostedtoolcache\windows\boost\1.72.0\x86_64\include\boost-1_72\boost/test/impl/compiler_log_formatter.ipp(64,1): warning C4273: 'boost::unit_test::output::compiler_log_formatter::log_start': inconsistent dll linkage [D:\a\1\s\build\test\connect.vcxproj]
```
https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1563&view=logs&j=77262b04-dc69-55ba-9da0-0c8477d74276&t=1b86177c-f53d-5127-4b2f-4b26b44c9d86&l=56

Is the latter error is the same as you experienced?

**Comment by kleunen** [comment]
> Before I define BOOST_TEST_SOURCE, I got the following errors:
> 
> ```
> C:\hostedtoolcache\windows\boost\1.72.0\x86_64\include\boost-1_72\boost/test/impl/compiler_log_formatter.ipp(64,1): warning C4273: 'boost::unit_test::output::compiler_log_formatter::log_start': inconsistent dll linkage [D:\a\1\s\build\test\connect.vcxproj]
> ```
> 
> https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1563&view=logs&j=77262b04-dc69-55ba-9da0-0c8477d74276&t=1b86177c-f53d-5127-4b2f-4b26b44c9d86&l=56
> 
> Is the latter error is the same as you experienced?

Yes, that is what I experience. I think it has to do that boost is linked dynamicly, but the test system assumes static linkage is used ?

**Comment by redboltz** [comment]
There is another problem.

Even if cmake is finished by failure, CI regard it to success.
Maybe my https://github.com/redboltz/mqtt_cpp/blob/master/azure-pipelines.yml notation is something wrong...

**Comment by kleunen** [comment]
> Maybe my https://github.com/redboltz/mqtt_cpp/blob/master/azure-pipelines.yml notation is something wrong...

This I really would not know

**Comment by redboltz** [comment]
> NOTE: I use azure build pipelines instead of GithubActions for MSVC because to overcome pallalellism limit.

Migrating from azure build pipelines to GithubActions could be a choice.
I'm not sure the problem would solve but worth trying if we completely stuck with azure build pipelines.

Here is a table when I checked the max number of parallelism before.

|Environment|Max number of parallelism|
|---|---|
|Linux|8|
|osx|5|
|Windows|?|

Perhaps the number would be updated.
I'm looking for the up-to-date value in Github Actions Document, but couldn't find, so far.

**Comment by kleunen** [comment]
Also that you need to define BOOST_TEST_SOURCE sounds like a hack to me, i could not find the original thread that mentioned this. But I think there is something wrong with using the unit test header

**Comment by redboltz** [comment]
I'm not sure but installing the boost library from source code might solve the problem on azure build pipelines.
But writing azure-pipelines.yml script for it is very hard for me ;) 
I'm not familiar with windows command line. And this approach might not work. 

Hmm...

**Comment by kleunen** [comment]
https://stackoverflow.com/questions/45880224/defining-boost-test-dyn-link-causes-application-to-crash-in-visual-studio

**Comment by kleunen** [comment]
I think you need to define BOOST_TEST_NO_LIB to disable auto library on windows, I am checking now

**Comment by kleunen** [comment]
The fix is to not define BOOST_TEST_DYN_LINK on MSVC:

````

FOREACH (source_file ${check_PROGRAMS})
    GET_FILENAME_COMPONENT (source_file_we ${source_file} NAME_WE)
    ADD_EXECUTABLE (${source_file_we} ${source_file})
    TARGET_LINK_LIBRARIES (
        ${source_file_we} mqtt_cpp_iface Boost::unit_test_framework
    )
    IF (MQTT_USE_LOG)
        TARGET_COMPILE_DEFINITIONS (${source_file_we} PUBLIC $<IF:$<BOOL:${MQTT_USE_STATIC_BOOST}>,,BOOST_LOG_DYN_LINK>)
        TARGET_LINK_LIBRARIES (
            ${source_file_we} Boost::log
        )
    ENDIF ()
    IF ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
        TARGET_COMPILE_DEFINITIONS (${source_file_we} PUBLIC $<IF:$<BOOL:${MQTT_USE_STATIC_BOOST}>,,BOOST_TEST_DYN_LINK>)

        IF (MQTT_CODECOV)
            SET_PROPERTY (TARGET ${source_file_we}
                          APPEND_STRING PROPERTY COMPILE_FLAGS " -O0 -g --coverage -fno-inline")
            SET_PROPERTY (TARGET ${source_file_we}
                          APPEND_STRING PROPERTY LINK_FLAGS " --coverage")
        ENDIF ()
    ENDIF ()
    IF ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
        IF (CMAKE_CXX_FLAGS MATCHES "/W[0-4]")
            STRING(REGEX REPLACE "/W[0-4]" "/W3" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
        ELSE ()
            SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W3")
        ENDIF ()
    ENDIF ()

    ADD_TEST (${source_file_we} ${source_file_we})
ENDFOREACH ()

**Comment by redboltz** [comment]
Thank you. I just applied your fix to #710

**Comment by redboltz** [comment]
Error message has been changed. But it is a great progress.
https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1584&view=logs&j=77262b04-dc69-55ba-9da0-0c8477d74276&t=1b86177c-f53d-5127-4b2f-4b26b44c9d86&l=156
I've not checked instantiate trace but the error happened in boost.

I will update boost 1.69.0 to 1.72.0.

**Comment by redboltz** [comment]
When I updated to 1.72.0. DLL related error is back.

https://redboltz.visualstudio.com/redboltz/_build/results?buildId=1586&view=logs&j=c9c4379a-274e-565b-38f2-6c517cc5236b&t=9b0c8de6-6118-5063-3ec7-612723e4a786&l=303

I don't know why...

**Comment by kleunen** [comment]
Maybe boost fixed themselves and now you do have to define dyn link?

**Comment by redboltz** [comment]
Just memo;
https://github.com/boostorg/thread/issues/264
I've check it later

**Comment by redboltz** [comment]
Fixed by #718
